framework/components/ADatePicker/useADateRange.js

```diff
@@ -0,0 +1,22 @@
+import { useState } from "react";
+  
+  const useADateRange = (initialRange  = [null, null]) => {
+    const [range, setRange] = useState(initialRange)
+  
+    const onChange = (date) => {
+      setRange(prevRange => {
+        const [prevStartDate, prevEndDate] = prevRange;
+        if (!prevStartDate && !prevEndDate) {
+          return [date, null];
+        }
+        return prevEndDate ? [date, null] : [prevStartDate, date];
+      })
+    }
```

Comment:
> Logic issue -> prevEndDate ? [date, null] : [prevStartDate, date]. Seems questionable. User have restart date selection if they clicked third time?

Yes, that was the idea. It's the same logic in [Vuetify's Date range picker component](https://vuetifyjs.com/en/components/date-pickers/#range). We can make the logic more complicated, but since there was no actual UX requirement, I wanted to implement something simpler in the case we don't actually need more sophisticated logic. This PR originated from #1062.

Reply:
I've refactored some of the logic by creating helper functions and renaming variables to hopefully improve some of the readability. I also wrapped the existing loop in a `useMemo` for a small performance boost.